### PR TITLE
feat: Add feature that allows merging Indicator Types [DHIS2-14458]

### PR DESCRIPTION
**Pull request by david-mackessy**
# Summary
Add feature that allows merging `IndicatorType`s

## Changes
- add new API endpoint `POST` `/api/indicatorTypes/merge`
- add generic structure to be able to process any metadata merge
  - `MergeParams`
  - `MergeRequest`
  - `MergeParamsProcessor`
  - `MergeService`
- add `F_INDICATOR_TYPE_MERGE` authority
- add `MergeReport` which is a lightweight report only containing `ErrorMessage`s
- Add `IndicatorType` merge error codes
- Add `@JsonCreator` annotation to `UID` class to allow receiving `String` `UID`s and converting them to the `UID` type (this adds better validation at the API front door)
- add new `@ExceptionHandler(HttpMessageNotReadableException.class)` to format a nice response for UID format errors

## Files renamed for existing merge service (after team discussion, prefix with metadata to distinguish between them)
- org.hisp.dhis.schema.MergeService -> org.hisp.dhis.schema.MetadataMergeService
- org.hisp.dhis.schema.DefaultMergeService -> org.hisp.dhis.schema.DefaultMetadataMergeService
- org.hisp.dhis.schema.MergeParams -> org.hisp.dhis.schema.MetadataMergeParams
- org.hisp.dhis.dxf2.metadata.MergeServiceTest -> org.hisp.dhis.dxf2.metadata.MetadataMergeServiceTest

# Testing
## Automated
- e2e tests
- service integration tests
- unit

## Manual
To test this feature, `Indicator`s and `IndicatorType`s should already exist.
To trigger a merge call:
- `POST` `/api/indicatorTypes/merge` with a payload like:
```json
{
    "sources": ["sourceUid1", "sourceUid2"],
    "target": "targetUid",
    "deleteSources": false
}
```

# Docs
docs PR: https://github.com/dhis2/dhis2-docs/pull/1260 

# Design
https://dhis2.atlassian.net/wiki/spaces/SOFTWARE/pages/176816151/Metadata+merge+design

### CODE REVIEW

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/indicator/IndicatorTypeController.java [hunk 1]
```diff
@@ -27,17 +27,53 @@
...
+  @PostMapping(value = "/merge", produces = APPLICATION_JSON_VALUE)
+  public @ResponseBody WebMessage mergeIndicatorTypes(@RequestBody MergeQuery query) {
+    log.info("Indicator type merge request received");
+
+    MergeReport report =
+        new MergeQueryProcessor<>(indicatorTypeMergeService)
+            .processMergeQuery(query, MergeType.INDICATOR_TYPE);
+
+    log.info("Indicator type merge request processed: {}", report);
+    return WebMessageUtils.mergeReport(report);
```
Comment (jbee) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/indicator/IndicatorTypeController.java:77]: I know this pattern of having a report handling method is used and established but it does hide possible error outcomes from OpenAPI. I think we should move towards having the `mergeReport` method throw a exception from the feedback package instead of directly generating different `WebMessage`s. Only the success message should be returned. The error cases can go via exceptions which in the end generate the same error web messages but in a way that is visible.

File: dhis-2/dhis-web-api/src/main/resources/openapi/AbstractCrudController.md [hunk 2]
```diff
@@ -52,7 +52,7 @@ Combine filters with `AND` (default) or `OR`
 ### `*.parameter.ImportOptions.preheatCache`
 Turn cache-map preheating on/off. This is on by default, turning this off will make initial load time for importer much shorter (but will make the import itself slower). This is mostly used for cases where you have a small XML/JSON file you want to import, and don't want to wait for cache-map preheating.
 
-### `*.parameter.ImportOptions.mergeMode`
+### `*.parameter.ImportOptions.metadataMergeMode`
```
Comment (jbee) [inline dhis-2/dhis-web-api/src/main/resources/openapi/AbstractCrudController.md:55]: This rename of the field needs to be reverted here and in the `ImportOptions` class. In general when renaming classes to add a prefix or suffix it generally is not a good idea to also rename all fields accordingly. The extra word is already in the type information so doubling it in the parameter or field name does only add clutter but now additional information. And as here it could break the API.

**Review by jbee — CHANGES_REQUESTED** [review]
Overall it looks really good. 

There are a couple of code style things, mostly code organisation and error handling stuff.
And some of the renames of fields/parameters need to be undone to not break public API.

File: dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/MergeQuery.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * {@link UID} to be merged in to. Also indicates whether sources should be deleted or not. <br>
+ * All {@link UID}s should be verified.
+ *
+ * @author david mackessy
+ */
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class MergeQuery {
+  @OpenApi.Property({UID[].class})
```
[the remark this answers is not in the document]
Comment (david-mackessy) [reply dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/MergeQuery.java:50]: cool, i'll remove this

File: dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/MergeRequest.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+ * Encapsulation of a merge request. Contains source types that extend {@link
+ * BaseIdentifiableObject} to be merged into a target of the same type. Also indicates whether
+ * sources should be deleted.
+ *
+ * @author david mackessy
+ */
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+public class MergeRequest<T extends BaseIdentifiableObject> {
```
[the remark this answers is not in the document]
Comment (david-mackessy) [reply dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/MergeRequest.java:50]: will update MergeQuery to MergeParams (use Lombok @Data)

File: dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/MergeQueryProcessor.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+ *   <li>transform
+ *   <li>validate
+ *   <li>merge
+ * </ol>
+ *
+ * @param <T> type extending BaseIdentifiableObject
+ */
+public class MergeQueryProcessor<T extends BaseIdentifiableObject> {
+
+  public MergeQueryProcessor(MergeService<T> mergeService) {
```
[the remark this answers is not in the document]
Comment (david-mackessy) [reply dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/MergeQueryProcessor.java:50]: will update

File: dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/MergeRequest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (david-mackessy) [reply dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/MergeRequest.java:50]: will try to update to use the UIDs and not the objects themselves.
Check if no args is required...

File: dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/MergeRequest.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  @Builder.Default private Set<T> sources = new HashSet<>();
+
+  private T target;
```
[the remark this answers is not in the document]
Comment (david-mackessy) [reply dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/MergeRequest.java:53]: will update

File: dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/MergeRequest.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  public static <T extends BaseIdentifiableObject> MergeRequest<T> empty() {
+    return MergeRequest.<T>builder().build();
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("sources", IdentifiableObjectUtils.getUids(sources))
```
[the remark this answers is not in the document]
Comment (david-mackessy) [reply dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/MergeRequest.java:72]: when change made to hold UIDs only, this will be safe to use with the new property

File: dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/MergeService.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+import org.hisp.dhis.feedback.MergeReport;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * Interface for merging {@link BaseIdentifiableObject}s
+ *
+ * @param <T> type extending {@link BaseIdentifiableObject}
+ * @author david mackessy
+ */
+public interface MergeService<T extends BaseIdentifiableObject> {
```
[the remark this answers is not in the document]
Comment (david-mackessy) [reply dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/MergeService.java:41]: will move this service interface into API module

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/indicator/IndicatorTypeController.java (hunk 1 above)
Comment (david-mackessy) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/indicator/IndicatorTypeController.java:77]: update possible thrown exceptions to include merge report, if merge report has errors then throw conflict exception

File: dhis-2/dhis-web-api/src/main/resources/openapi/AbstractCrudController.md (hunk 2 above)
Comment (david-mackessy) [reply dhis-2/dhis-web-api/src/main/resources/openapi/AbstractCrudController.md:55]: will revert this

Comment (david-mackessy) [reply dhis-2/dhis-web-api/src/main/resources/openapi/AbstractCrudController.md:55]: will leave it as just class renames - will revert variable names

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/merge/MergeParams.java [hunk 9]
```diff
@@ -30,30 +30,22 @@
...
-  private Set<UID> sources = new HashSet<>();
+public class MergeParams {
+  @JsonProperty private final Set<UID> sources = new HashSet<>();
 
-  @OpenApi.Property(UID.class)
-  @JsonProperty
-  private UID target;
+  @JsonProperty private UID target;
 
   @OpenApi.Property @JsonProperty private boolean deleteSources;
```
[the remark this answers is not in the document]
Comment (david-mackessy) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/merge/MergeParams.java]: that was removed a while ago. Just going through some changes still.

**Comment by david-mackessy** [comment]
> Overall it looks really good.
> 
> There are a couple of code style things, mostly code organisation and error handling stuff. And some of the renames of fields/parameters need to be undone to not break public API.

@jbee comments addressed now & ready for re-review

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/merge/MergeParamsProcessor.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+ * in a standardised fashion. It requires an implementation of {@link MergeService}, which will
+ * process the merging for its required use case. <br>
+ * It essentially calls each method of the {@link MergeService} in the following order:<br>
+ *
+ * <ol>
+ *   <li>validate
+ *   <li>merge
+ * </ol>
+ */
+public interface MergeParamsProcessor {
```
Comment (jbee) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/merge/MergeParamsProcessor.java:46]: I think the naming is slightly misleading here. This is not about params but the merge itself. Therefore I would call this just `MergeProcessor` and the method `processMerge`.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/merge/MergeRequest.java [hunk 11]
```diff
@@ -33,44 +33,37 @@
...
+    return MergeRequest.builder().build();
   }
 
   @Override
   public String toString() {
     return MoreObjects.toStringHelper(this)
-        .add("sources", IdentifiableObjectUtils.getUids(sources))
-        .add("target", target != null ? target.getUid() : null)
+        .add("sources", sources)
+        .add("target", target)
```
[the remark this answers is not in the document]
Comment (jbee) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/merge/MergeRequest.java:67]: :point_up:

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/indicator/IndicatorTypeController.java [hunk 12]
```diff
@@ -27,17 +27,56 @@
...
+  @ResponseStatus(HttpStatus.OK)
+  @PreAuthorize("hasRole('ALL') or hasRole('F_INDICATOR_TYPE_MERGE')")
+  @PostMapping(value = "/merge", produces = APPLICATION_JSON_VALUE)
+  public @ResponseBody WebMessage mergeIndicatorTypes(@RequestBody MergeParams params)
+      throws ConflictException {
+    log.info("Indicator type merge received");
+
+    MergeReport report =
+        mergeParamsProcessor.processMergeParams(
+            indicatorTypeMergeService, params, MergeType.INDICATOR_TYPE);
```
Comment (jbee) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/indicator/IndicatorTypeController.java:77]: The `IndicatorTypeMergeParamsProcessor` is a spring component so the service should not be passed as a parameter here but be injected by spring when constructing the component.

**Review by jbee — APPROVED** [review]
Almost there :D 
A few minor recommendations.

For my taste the service implementation is relying on `Optional` too much making it hard to follow the control flow logic as it is mainly build into the optional transformations.

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/indicator/IndicatorTypeController.java (hunk 12 above)
Comment (david-mackessy) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/indicator/IndicatorTypeController.java]: rejigged wiring

**Comment by david-mackessy** [comment]
> Almost there :D A few minor recommendations.
> 
> For my taste the service implementation is relying on `Optional` too much making it hard to follow the control flow logic as it is mainly build into the optional transformations.

Yep, this is a style difference. Either use 'if else' checks or optional and subsequent processing. Either approach read very similar and are interchangeable really. Using either one shouldn't make a difference.

**Review by vietnguyen — APPROVED** [review]
